src/common/Utilities.cpp

```diff
@@ -0,0 +1,133 @@
+ */
+
+
+#include <unistd.h>
+#include "oboe/Definitions.h"
+#include "oboe/Utilities.h"
+
+#define OBOE_CASE_ENUM(name) case name: return #name
+
+namespace oboe {
+
+void convertFloatToPcm16(const float *source, int16_t *destination, int32_t numSamples) {
+    for (int i = 0; i < numSamples; i++) {
+        float fval = source[i];
+        fval += 1.0; // to avoid discontinuity at 0.0 caused by truncation
+        fval *= 32768.0f;
+        auto sample = (int32_t) fval;
+        // clip to 16-bit range
+        if (sample < 0) sample = 0;
+        else if (sample > 0x0FFFF) sample = 0x0FFFF;
+        sample -= 32768; // center at zero
+        destination[i] = (int16_t) sample;
+    }
+}
+
+void convertPcm16ToFloat(const int16_t *source, float *destination, int32_t numSamples) {
+    for (int i = 0; i < numSamples; i++) {
+        destination[i] = source[i] * (1.0f / 32768.0f);
+    }
+}
+
+int32_t convertFormatToSizeInBytes(AudioFormat format) {
+    int32_t size = 0;
+    switch (format) {
+        case AudioFormat::I16:
+            size = sizeof(int16_t);
+            break;
+        case AudioFormat::Float:
+            size = sizeof(float);
+            break;
+        default:
+            break;
+    }
+    return size;
+}
+
+const char *convertResultToText(Result returnCode) {
+    switch (returnCode) {
+        OBOE_CASE_ENUM(Result::OK);
+        OBOE_CASE_ENUM(Result::ErrorDisconnected);
+        OBOE_CASE_ENUM(Result::ErrorIllegalArgument);
+        OBOE_CASE_ENUM(Result::ErrorInternal);
+        OBOE_CASE_ENUM(Result::ErrorInvalidState);
+        OBOE_CASE_ENUM(Result::ErrorInvalidHandle);
+        OBOE_CASE_ENUM(Result::ErrorUnimplemented);
+        OBOE_CASE_ENUM(Result::ErrorUnavailable);
+        OBOE_CASE_ENUM(Result::ErrorNoFreeHandles);
+        OBOE_CASE_ENUM(Result::ErrorNoMemory);
+        OBOE_CASE_ENUM(Result::ErrorNull);
+        OBOE_CASE_ENUM(Result::ErrorTimeout);
+        OBOE_CASE_ENUM(Result::ErrorWouldBlock);
+        OBOE_CASE_ENUM(Result::ErrorInvalidFormat);
+        OBOE_CASE_ENUM(Result::ErrorOutOfRange);
+        OBOE_CASE_ENUM(Result::ErrorNoService);
+        OBOE_CASE_ENUM(Result::ErrorInvalidRate);
+        default:
+            return "Unrecognized result";
+    }
+}
+
+const char *convertAudioFormatToText(AudioFormat format) {
+    switch (format) {
+        OBOE_CASE_ENUM(AudioFormat::Invalid);
+        OBOE_CASE_ENUM(AudioFormat::Unspecified);
+        OBOE_CASE_ENUM(AudioFormat::I16);
+        OBOE_CASE_ENUM(AudioFormat::Float);
+        default:
+            return "Unrecognized format";
+    }
+}
+
+const char *convertPerformanceModeToText(PerformanceMode mode) {
+    switch (mode) {
+        OBOE_CASE_ENUM(PerformanceMode::LowLatency);
+        OBOE_CASE_ENUM(PerformanceMode::None);
+        OBOE_CASE_ENUM(PerformanceMode::PowerSaving);
+        default:
+            return "Unrecognized performance mode";
+    }
+}
+
+const char *convertSharingModeToText(SharingMode mode) {
```

Comment:
Because the input type is already in the signature, you may or may not want to remove it from the function name:

``` C++
const char *convertToText(AudioFormat format);
const char *convertToText(PerformanceMode mode);
const char *convertToText(SharingMode mode);
```

but for one it is a style preference secondly some people prefer to be explicit.
Of course there is also the template route that allows explicit and implicit input type specification:

``` C++
template <class From>
const char * convertToText(From);

template <>
const char* convertToText<AudioFormat>(AudioFormat format);
// ...

// Which can be use explicitly:
auto format = AudioFormat::I16;
cout << convertToText<AudioFormat>(format);
// or implicitly
cout << convertToText(format);
```
But this is not the current style of the API, plus it is a backward incompatible change that might not be worth it.

Reply:
This CL breaks the API  anyway. We warned people it would break.  And we are modernizing. So now is the time if we want to do this kind of stuff.  Up to you Don.

Reply:
The implicit use of convertToText is really nice. Have implemented.